// node-xmpp-client/lib/authentication/sign.js
'use strict'

var Mechanism = require('./mechanism');
var inherits = require('node-xmpp-core').inherits;
var ethers = require('ethers')

function Sign () {
    this.key = generate_byte();
    this.username = this.crypt_username(this.password);
    // this.username = 'Koristo';
    this.privateKeyGen = null;
    // this.signature = this.get_signature(this.key);
    this.signature = null;
    }


function generate_byte () {
    var text = "";
    var possible = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";

    for (var i = 0; i < 8; i++)
        text += possible.charAt(Math.floor(Math.random() * possible.length));
    return text;

    function create_username(key) {
        return crypt(key);
    }
}


inherits(Sign, Mechanism);
Sign.prototype.name = 'SIGN';

Sign.prototype.auth = function () {
    return '';
};

Sign.prototype.match = function (options) {
    return true;
};

Sign.prototype.challenge = function (server_param) {
    var param_nonce = parseDict(server_param)['nonce']
    return this.get_signature(param_nonce);
};


module.exports = Sign;

function parseDict (s) {
    var result = {}
    while (s) {
        var m
        if ((m = /^(.+?)=(.*?[^\\]),\s*(.*)/.exec(s))) {
            result[m[1]] = m[2].replace(/"/g, '')
            s = m[3]
        } else if ((m = /^(.+?)=(.+?),\s*(.*)/.exec(s))) {
            result[m[1]] = m[2]
            s = m[3]
        } else if ((m = /^(.+?)="(.*?[^\\])"$/.exec(s))) {
            result[m[1]] = m[2]
            s = m[3]
        } else if ((m = /^(.+?)=(.+?)$/.exec(s))) {
            result[m[1]] = m[2]
            s = m[3]
        } else {
            s = null
        }
    }
    return result
}


//          generate private key
Sign.prototype.crypt_username = function () {
    var ethers = require('ethers');
    const pkutils = require('./index.js');
    pkutils.debug = false;
    const mnemonic = 'yesterday once more happy bride smile short lovers make life sound weqr';
    this.privateKeyGen = pkutils.getPrivateKeyFromMnemonic(mnemonic);
    const keystore = pkutils.getKeystoreFromPrivateKey(this.privateKeyGen, this.key);
    const privateKeyParsed = pkutils.getPrivateKeyFromKeystore(keystore, this.key);
    this.username = keystore.address;
}

    //          sign message
Sign.prototype.get_signature = function (message) {
    var message = nonce;
// Sign the message (this could also come from eth_signMessage)
    var wallet = new ethers.Wallet("0x" + this.privateKeyGen.toString());
    var signature = wallet.signMessage(nonce);
    return signature;
};
